### ISSUE TITLE
Ditto doesn't reply to Kafka topic when multiple commands are sent

### ISSUE THREAD

**Issue by hfcr2017** [eclipse/ditto#1458]
I'm currently running a test where multiple messages are sent to Ditto and I expected the responses back. It works for around 70 - 80% of messages that I sent, but some messages are just missing.

The integration is as follows:

1. I put multiple messages in a Kafka topic, namely, `main.ditto.topic`.
2. The messages are fetched by a connection that I have configured in Ditto.
3. Ditto should send the response to a different Kafka topic, namely, `response.ditto.topic`.

Here's how I have currently configured the connection:

```
{
    "connectivity": {
        "ditto-connectivity": {
            "type": "connectivity.responses:retrieveConnection",
            "status": 200,
            "connection": {
                "id": "my-connector",
                "name": null,
                "connectionType": "kafka",
                "connectionStatus": "open",
                "uri": "my.uri",
                "sources": [
                    {
                        "addresses": [
                            "main.ditto.topic"
                        ],
                        "consumerCount": 1,
                        "qos": 1,
                        "authorizationContext": [
                            "my:auth:context"
                        ],
                        "headerMapping": {
                            "reply-to": "response.ditto.topic",
                            "response-required": "true",
                            "timeout": "1800s"
                        },
                        "payloadMapping": [
                            "Ditto"
                        ],
                        "replyTarget": {
                            "address": "response.ditto.topic",
                            "headerMapping": {},
                            "expectedResponseTypes": [
                                "response",
                                "error",
                                "nack"
                            ],
                            "enabled": true
                        }
                    }
                ],
                "targets": [],
                "clientCount": 1,
                "failoverEnabled": true,
                "validateCertificates": true,
                "processorPoolSize": 1,
                "specificConfig": {
                    "saslMechanism": "plain",
                    "bootstrapServers": "bootstrap-server"
                },
                "tags": []
            }
        }
    }
}
```

I have tried other many configurations: adding `twin-persisted`, removing `reply-to`, etc. But none of them have made any difference.

The messages that I have in `main.ditto.topic` are `modify` commands, like this:

```
{
  "path": "/the/path",
  "headers": {
    "correlation-id": "the-correlation-id"
  },
  "topic": "namespace/mything/things/twin/commands/modify",
  "value": [
    {
      "my-load": "OK"
    }
  ]
}
```

So I expected to see some kind of response (as it happens for other things. I can see some information logged in `ditto-connectivity`, but there is nothing on the `response.ditto.topic`. Here's what I see in the logs:

```
DEBUG [c47e593d-52ce-49f9-a5ad-9ce7b51bbdbc] o.e.d.c.s.m.InboundMappingSink  - Handling ExternalMessage: UnmodifiableExternalMessage
```
and 

```
DEBUG [] o.e.d.c.s.m.InboundDispatchingSink  - OnMapped from <Actor[akka://ditto-cluster/system/sharding/connection/8/ddm-thing-fetcher/pa/$M/c1/kafkaConsumer-0-main.ditto.topic-02/$7q#2140000697]>: <InboundMappingOutcomes [outcomes=[MappedOutcome[mapperId=Ditto,mapped=MappedInboundExternalMessage [externalMessage=UnmodifiableExternalMessage
```

However, I don't see anything related to the response. It looks like Ditto didn't try to reply at all.

**Comment by danielsularea** [comment]
+1 I am also experiencing a similar issue.

**Comment by Yannic92** [comment]
Hi @hfcr2017 ,

Thanks for reaching out. Before starting to dig into the problem I would like to make sure I got everything right.

> It works for around 70 - 80% of messages that I sent, but some messages are just missing.

Do those 70-80% relate to your described use case? Or are there other messages included?
Im asking because of

> It looks like Ditto didn't try to reply at all.

Is this meant for a single message that did not receive a response? Or for the described use case as a whole?

Also: have you tried to [enable connection logging](https://www.eclipse.org/ditto/connectivity-manage-connections.html#enable-connection-logs) and [retrieve them](https://www.eclipse.org/ditto/connectivity-manage-connections.html#retrieve-connection-logs) after you ran your test?

The [connection metrics](https://www.eclipse.org/ditto/connectivity-manage-connections.html#retrieve-connection-logs) could also be of interest to track this issue down.

**Comment by hfcr2017** [comment]
Hi @Yannic92 ,

Thanks for your reply. I have updated the connection to use `60s` as timeout. I also have some new findings.

Regarding

> Do those 70-80% relate to your described use case? Or are there other messages included?

I can add some more details to the type of test that I'm doing. So we send, let's say, 10000 `modify` commands to Ditto in a given point of time. Each one has its own identifier. They are read by the connector and in case that the thing doesn't exist, we expect that Ditto will send a 404 response message to `response.ditto.topic`. So in the worst case where none of the things can be found, we expect to receive 10000 responses with 404 response code. However, we receive only 70 to 80% of these messages. That's 7000 - 8000 messages. For the rest, there is no response at all in the Kafka topic.

I have enabled the connection logging but there is not much of information, as given the amount of messages, the logs retrieved are just a small amount.

Now, I have increased the amount of `ditto-connectivity` instances (to 6) and I can see a general improvement. I also increased the `clientCount` to 6 in the connection configuration. I got 100% response messages. So now my feeling is that Ditto drops some messages when it can't handle them. What do you think? How can I prove this point?

Moreover, is there any relation that you might already know between the amount of instances that I need given an amount of messages?

**Comment by thjaeckle** [comment]
Ah, that kind of information was missing for us :)

Yes, Ditto (must) drops messages when we can not get them quickly to Kafka, see the config:
https://github.com/eclipse/ditto/blob/fc6f942b8c0c6df1ffb903116f9dc45e9c613829/connectivity/service/src/main/resources/connectivity.conf#L545-L552

If Ditto wouldn't drop, it would at some point crash with out-of-memory errors.

However, to mitigate that, you can configure the "consumer" to throttle when receiving messages from Kafka (which is deactivated by default):
https://github.com/eclipse/ditto/blob/fc6f942b8c0c6df1ffb903116f9dc45e9c613829/connectivity/service/src/main/resources/connectivity.conf#L436-L462

Or you can scale out via "clientCount" (like you did) - or you can increase the "producer" queue size how much messages Ditto will queue.

I am however wondering: Is this a real scenario?
That a Kafka already has 10.000 messages persisted which Ditto should process when connecting to it?
Or do you just use that to easily simulate a load test? Because that is not a realistic scenario, or is it? Messages are processed all the time and such "huge" batches should not occur.
If they do, then yes .. Ditto must be probably scaled as well.

> Moreover, is there any relation that you might already know between the amount of instances that I need given an amount of messages?

That depends on so many factors (message size, involved optional payload mapping, assigned CPUs of the connectivity instances, performance of the network, scaling of other Ditto services, performance of the database, etc.).
So: "it depends". ;)

**Comment by hfcr2017** [comment]
Hi @thjaeckle,

Thanks a lot for the information! To answer your question 

> I am however wondering: Is this a real scenario?

Yeah, it is a possible scenario that we might face when processing updates from thousands of devices that will come in a batch.

I will try enabling the `throttling` and will come back to you as soon as possible.

**Comment by hfcr2017** [comment]
Hi @thjaeckle,

The `throttling` works good and it has helped us to avoid that messages get lost, however, we have struggled quite a lot with setting the correct values of KAFKA_CONSUMER_THROTTLING_LIMIT, KAFKA_PRODUCER_QUEUE_SIZE and the number of `ditto-connectivity` instances that we need to have. I understand that there is no other way to configure this parameters but with manual tuning, is that correct?

Now, we have also faced some issue where I can't find things if I use the `/search/things` API but I do find them if a use the `thing` API with the specific ID. Why is this even possible?

**Comment by thjaeckle** [comment]
> I understand that there is no other way to configure this parameters but with manual tuning, is that correct?

Yes, correct. Do you have an idea how an automatic tuning of these parameters could look like? We're always open for improvement suggestions. 

> Now, we have also faced some issue where I can't find things if I use the `/search/things` API but I do find them if I use the `thing` API with the specific ID. Why is this even possible? 

This is possible because the search needs its own database / Mongo collection to put a search index over all the things, including their permissions based on the used policies.
The search provides only eventual consistency. 

In the upcoming Ditto 3.0 this search index will massively improve in both performance and consistency.
Also fixing bugs in Ditto 2.x regarding skipped search index updates.
As this new search in Ditto 3.0 also required to change the database schema, backporting to Ditto 2.x is unfortunately not possible.

Ditto 3.x will be released this month.

**Comment by Yannic92** [comment]
Since the original question about missing responses seems to be solved , I close this issue.
If you (@hfcr2017) or someone else is interested in a discussion about the improvement of ditto config tuning, this would be worth another issue. Feel free to raise a new issue regarding the topic.
